packages/liveblocks-core/src/__tests__/_utils.ts

```diff
@@ -50,94 +56,114 @@ function deepClone<T extends Json>(items: T): T {
 
-  static instances: MockWebSocket[] = [];
+export class MockWebSocket {
+  readonly CONNECTING = 0;
+  readonly OPEN = 1;
+  readonly CLOSING = 2;
+  readonly CLOSED = 3;
 
-  isMock = true;
+  readonly url: string;
 
-  callbacks = {
-    open: [] as Array<(event?: WebSocketEventMap["open"]) => void>,
-    close: [] as Array<(event?: WebSocketEventMap["close"]) => void>,
-    error: [] as Array<(event?: WebSocketEventMap["error"]) => void>,
-    message: [] as Array<(event?: WebSocketEventMap["message"]) => void>,
-  };
+  #readyState: number;
+  readonly sentMessages: string[] = [];
 
-  sentMessages: string[] = [];
-  readyState: number;
+  readonly #closeListeners: CloseListener[] = [];
+  readonly #errorListeners: Listener[] = [];
+  readonly #messageListeners: MessageListener[] = [];
+  readonly #openListeners: Listener[] = [];
 
-  constructor(
-    public url: string,
-    private onSend: (message: string) => void = () => {}
-  ) {
-    this.readyState = this.CONNECTING;
-    MockWebSocket.instances.push(this);
+  constructor(url: string = "ws://ignored") {
+    this.url = url;
+    this.#readyState = this.CONNECTING;
   }
 
-  close(_code?: number, _reason?: string): void {
-    this.readyState = this.CLOSED;
+  //
+  // WEBSOCKET API
+  //
+
+  get readyState(): number {
+    return this.#readyState;
   }
 
-  addEventListener<T extends "open" | "close" | "error" | "message">(
-    event: T,
-    callback: (event: WebSocketEventMap[T]) => void
-  ) {
-    this.callbacks[event].push(callback as any);
+  addEventListener(event: "message", listener: MessageListener): void; // prettier-ignore
+  addEventListener(event: "close", listener: CloseListener): void; // prettier-ignore
+  addEventListener(event: "open" | "error", listener: Listener): void; // prettier-ignore
+  // prettier-ignore
+  addEventListener(event: "open" | "close" | "message" | "error", listener: Listener | MessageListener | CloseListener): void {
+    if (event === "open") {
+      this.#openListeners.push(listener as Listener);
+    } else if (event === "close") {
+      this.#closeListeners.push(listener as CloseListener);
+    } else if (event === "error") {
+      this.#errorListeners.push(listener as Listener);
+    } else if (event === "message") {
+      this.#messageListeners.push(listener as MessageListener);
+    }
   }
 
-  removeEventListener<T extends "open" | "close" | "error" | "message">(
-    event: T,
-    callback: (event: WebSocketEventMap[T]) => void
-  ) {
-    remove(this.callbacks[event], callback as any);
+  removeEventListener(event: "message", listener: MessageListener): void; // prettier-ignore
+  removeEventListener(event: "close", listener: CloseListener): void; // prettier-ignore
+  removeEventListener(event: "open" | "error", listener: Listener): void; // prettier-ignore
+  // prettier-ignore
+  removeEventListener(event: "open" | "close" | "message" | "error", listener: Listener | MessageListener | CloseListener): void {
```

Comment:
maybe consider making an event type rather than a list of strings.

Reply:
I had that first, but went back this way with string literals to more clearly see from the outside which event names exist without having to look up those names. TypeScript will protect from typos in these string literals, so the duplication is fine here I think.